test/test_gpu3.cpp

```diff
@@ -5386,13 +5386,6 @@ TEST_F(NVFuserTest, AsyncCompilation_CUDA) {
 
   std::vector<c10::IValue> aten_inputs = {t0, t1, t2};
 
-  executor_cache.compileFusionAsync(aten_inputs);
```

Comment:
How is the parallel compilation used? Is this test still "async"?

Reply:
The different segments of a fusion are compiled in parallel. Compilation is synchronized at the end of `compileFusionParallel`.

Reply:
So, this is automatically done always? Nice!

Reply:
Is there any way to adjust the parallelism of compilations? Something like `OMP_NUM_THREADS`.

Reply:
I've set it arbitrarily to 24 threads [here](https://github.com/NVIDIA/Fuser/blob/7b95d846057413af0ab2ab953eebe96da7b25426/csrc/kernel_cache.cpp#L28).

We could set it to the max available threads: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency

Or did you want a controllable knob in `cmake`?

Reply:
Yeah, but not compile time but some runtime knob.

Also, 24 seems quite large to me, although I don't think having that many segments would be quite rare. Why not make the default a small number, say, 4, and make it possible to adjust with some environment variable?

Reply:
Generally speaking, since nvfuser is a library embedded in user applications, we should not be too aggressive about usage of finite resources

Reply:
I'm assuming here a thread is an OS thread, not like a user-level lightweight thread. Not very familiar with the C++ threading package.

Reply:
Just FYI, using more threads does help with running the cpp tests faster because we have fusions with several segments.
* FusionCastings_CUDA had 81 kernels.
* FusionStandaloneFull_CUDA had 27 kernels.
My CPU usage didn't go about 20% when using max available threads, but it helped compile time latency.

Reply:
Instead of manual opt-in, should we just use this?
https://github.com/pytorch/pytorch/blob/07e759eca240730d8cb8e1f498fc4aa526733a47/aten/src/ATen/Parallel.h#L19C1-L20

Reply:
I'm not sure how thread parallelism is managed in PyTorch, but how do you know how many threads out of the number returned by the API? I'm assuming that should give the maximum, so no other things are using any additional threads, maybe it's fine to use that number of threads, but if there are CPU threads computing some ops in parallel, maybe we should limit our thread usage.

I don't think we need to build out something significant at this point. We should just start more conservatively.

Reply:
I think this number comes from https://pytorch.org/docs/stable/generated/torch.set_num_threads.html?highlight=set_num_threads, which is "number of threads used for intraop parallelism on CPU", do we consider each CUDA fusion group as an "op"? If so, I think it should be safe to use this "intraop" thread count? We will have threads competing for resources if we run multiple CPU ops in parallel though.

Reply:
For example, in a cpu reduction kernel: https://github.com/pytorch/pytorch/blob/326a4cc8156a4d9eb11bb4ef2a275f81bfe96742/aten/src/ATen/native/TensorIteratorReduce.cpp#L42
My understanding is, in PyTorch, each op just get `get_num_threads` and run on this number of threads, without caring about others.

Reply:
And `get_num_threads` just calls `omp_get_max_threads`
https://github.com/pytorch/pytorch/blob/326a4cc8156a4d9eb11bb4ef2a275f81bfe96742/aten/src/ATen/ParallelOpenMP.cpp#L73-L84

Reply:
I'm not a PyTorch user, but to me it doesn't sound quite right to consider CUDA kernel compilation is "intraop parallelism on CPU".

Reply:
I added a `NVFUSER_NUM_TRHEADS` environment variable and set the default number to 4.

Reply:
> I added a `NVFUSER_NUM_TRHEADS` environment variable and set the default number to 4.

Then @xwang233 should we set this to a larger number in our internal CI to reduce CI time?

Reply:
> > I added a `NVFUSER_NUM_TRHEADS` environment variable and set the default number to 4.
> 
> Then @xwang233 should we set this to a larger number in our internal CI to reduce CI time?

I think we should also do that `NVFuserTest::SetUp` as well as in the Python test scripts.

Reply:
Latest local CI running time:
- 1 thread - (637,408 ms total)
- 4 threads - (613,087 ms total)
- 24 threads - (594,463 ms total)

I think we would get most of the benefit if the default was set to 8.

Reply:
That sounds reasonable too.

Reply:
> To me, what's most important is to provide a way to control the parallelism. It seems Ryan already did that.

To provide a way to control, I totally agree. I am more arguing about the default value. And I don't think the discussion about default value should block the PR.

> Well, even if something else is not a good citizen, I don't think it would automatically mean being a good citizen wouldn't matter at all.

I don't think PyTorch is not a "good citizen" and I think this assumption does make sense. So I am kind of arguing, I think this is a reasonable default value, and we should follow it.

> Am I too conservative?

I think you prioritize more about making nobody unhappy, but I am more at if there is a change that will make 51% user happy and 49% user unhappy, then I will say yes to that change. I do think that using `get_num_threads` as default value is such a change. On the other hand, I don't think there is a clear answer about which philosophy is better, and I don't think it worth much time to discuss it. So whatever Ryan choose, that's the answer.